aidial_assistant/json_stream/json_node.py

```diff
@@ -33,22 +37,38 @@ def char_position(self) -> int:
         return self._char_position
 
     @abstractmethod
-    def value(self) -> T:
+    def value(self) -> TValue:
         pass
 
 
-class ComplexNode(JsonNode[T], ABC, Generic[T]):
-    def __init__(self, char_position: int):
+class ReadableNode(
+    JsonNode[TValue], AsyncIterator[TElement], ABC, Generic[TValue, TElement]
+):
+    def __init__(self, source: AsyncIterator[TElement], char_position: int):
         super().__init__(char_position)
+        self._source = source
+
+    @override
+    def __aiter__(self) -> AsyncIterator[TElement]:
+        return self
+
+    @override
+    async def __anext__(self) -> TElement:
+        result = await anext(self._source)
+        self._accumulate(result)
+
+        return result
 
     @abstractmethod
-    async def parse(
-        self, stream: CharacterStream, dependency_resolver: NodeResolver
-    ):
+    def _accumulate(self, element: TElement):
         pass
 
+    async def read_to_end(self):
+        async for _ in self:
+            pass
+
 
-class PrimitiveNode(JsonNode[T], ABC, Generic[T]):
+class PrimitiveNode(JsonNode[TValue], ABC, Generic[TValue]):
```

Comment:
From the names of the classes it's not clear that they are mitually exclusive.
I suggest to rename:

`PritimiveNode` -> `AtomicNode`

`ReadableNode` -> `CompoundNode`